posts/MarkoB/2017-02-06-validator-service-loader.adoc

```diff
@@ -0,0 +1,290 @@
+= "Using ServiceLoader to add custom constraint definitions"
+Marko Bekhta
+:awestruct-tags: [ "Hibernate Validator", "Discussions" ]
+:awestruct-layout: blog-post
+---
+
+Have you ever been in a situation, when you would like to get even more from the lib
+that you are using all the time? Or when you want to somehow change its behavior ?
+
+Today we'll be talking about Hibernate Validator and how you can provide your own constraint
+and/or validator in a fully self-contained manner. Meaning packaging it all into it's own jar file,
+in a way that the others can use your lib for their needs.
+
+This functionality is based on Hibernate Validators usage of Java's https://docs.oracle.com/javase/8/docs/api/java/util/ServiceLoader.html[ServiceLoader]
+mechanism that allows to register additional constraint definitions
+(https://docs.jboss.org/hibernate/stable/validator/reference/en-US/html_single/#_constraint_definitions_via_code_serviceloader_code[see here]).
+
+How does it work? It's really simple, all you need is to provide your own constraint definitions (ServiceProviders)
+and register them in a specific file in `META-INF/services` directory of your jar. But more on that later.
+
+== Use standard annotations for nonstandard classes
+
+First let's consider a case when you would want one (or few) of standard bean validation
+annotations (or not as standard Hibernate Validator ones) to support some other type, besides
+the ones that are already supported.
+
+=== ThreeTen Extra types validation
+
+So what can be a real life scenario for building your own lib with constraints and sharing it? Well let's say that
+you are building some library with data classes that user might want to validate. As it would be almost
+impossible to keep track of all such libraries and write/maintain all those constraints for them - Hibernate
+Validator provides authors of such libs a possibility to write and share their own validation extensions.
+Which can be picked up by HV and used to validate your data classes.
+
+In this section we will look at http://www.threeten.org/threeten-extra/[ThreeTen Extra types] - a great lib
+that provides additional date-time classes to complement those already present in Java.
+
+As you may already know, Bean Validation (and Hibernate Validator in particular) provides support for the range of
+the https://jcp.org/en/jsr/detail?id=310[JSR 310] date types (javax.time API), by the use of
+`@Past`/`@Future` annotations. So we would want to use these annotations on ThreeTen Extra types as well.
+
+To still keep this example simple we will provide validators only for http://www.threeten.org/threeten-extra/apidocs/org/threeten/extra/DayOfMonth.html[`DayOfMonth`]
+and http://www.threeten.org/threeten-extra/apidocs/org/threeten/extra/Quarter.html[`Quarter`].
+
+We will create a maven project with two modules - one will contain validators, another will contain tests.
```

Comment:
the idea here was to have a module with validators which represent a "shared with the world" jar file. And the other one to be a "consumer" of "shared with the world" library. 
So in other words that we don't want to test the validator itself, if it correctly validates an object (which as you've said should be separated) but show that the built lib can be used in other projects (in our case - just other modules)

Don't know if that cleared it :)